### PR TITLE
Implement GK sketching on GPU.

### PR DESCRIPTION
**Pull request by trivialfis**
* The whole sketching pipeline is available on GPU.  Sorted entries are no longer copied back to host.
* Now GPU is able to load extremely wide and sparse datasets like URL.
* Bottleneck on Device Quantile DMatrix is now on Ellpack construction instead of sketching.  But overall performance is quite good now.
* Unifying some functions between sketching on DMatrix and Adapter.
* Now we need to track memory usage on the whole pipeline, which is tricky but still manageable.  Before having an inplace sketching algorithm, I couldn't think of an alternative.  Memory usage calculation is unified between test and real algorithm.
* Tests for GK are added.  The distributed test requires dask to setup the environment then run gtest.  This way we have a finer test to compare the result.

### CODE REVIEW

File: src/common/hist_util.cu [hunk 1]
```diff
@@ -109,39 +98,81 @@ void ExtractWeightedCuts(int device,
...
+  auto result = std::min(nnz, if_dense);
+  return result;
+}
 
-  dh::caching_device_vector<SketchEntry> cuts(num_columns * num_cuts);
-  ExtractCuts(device, num_cuts,
-              dh::ToSpan(sorted_entries),
-              dh::ToSpan(column_sizes_scan),
-              dh::ToSpan(cuts));
+size_t RequiredMemory(bst_row_t num_rows, bst_feature_t num_columns, size_t nnz,
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/common/hist_util.cu:126]: Not sure.  Best option would be inplace sketching.  But I think we came to the conclusion that it's not possible at this moment.

File: src/common/host_device_vector.cu [hunk 2]
```diff
@@ -205,10 +205,10 @@ class HostDeviceVectorImpl {
     // data is on the host
     LazyResizeDevice(data_h_.size());
     SetDevice();
-    dh::safe_cuda(cudaMemcpy(data_d_->data().get(),
-                             data_h_.data(),
-                             data_d_->size() * sizeof(T),
-                             cudaMemcpyHostToDevice));
+    dh::safe_cuda(cudaMemcpyAsync(data_d_->data().get(),
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/common/host_device_vector.cu:208]: It's copying to device.

File: src/common/quantile.cu [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+                   h_out_columns_ptr.begin());
+  auto d_out_columns_ptr = p_cuts->cut_ptrs_.ConstDeviceSpan();
+
+  // Set up output cuts
+  size_t total_bins = h_out_columns_ptr.back();
+  p_cuts->cut_values_.SetDevice(device_);
+  p_cuts->cut_values_.Resize(total_bins);
+  auto out_cut_values = p_cuts->cut_values_.DeviceSpan();
+
+  dh::LaunchN(0, total_bins, [=] __device__(size_t idx) {
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/common/quantile.cu:529]: There are some heuristics jammed in creating the final cuts.  Maybe I need to explain it more with comments.

**Comment by trivialfis** [comment]
> It seems you accidentally changed the permissions for some files.

I corrected them.  It's very distracting during development ...

File: src/common/quantile.cu (hunk 3 above)
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/common/quantile.cu:529]: Added some comments.

File: src/common/device_helpers.cuh [hunk 4]
```diff
@@ -886,9 +978,85 @@ DEV_INLINE void AtomicAddGpair(OutputGradientT* dest,
...
+  using Key = thrust::pair<size_t, typename thrust::iterator_traits<ValInIt>::value_type>;
+  dh::XGBCachingDeviceAllocator<char> alloc;
+  auto unique_key_it = dh::MakeTransformIterator<Key>(
+      thrust::make_counting_iterator(static_cast<size_t>(0)),
+      [=] __device__(size_t i) {
+        size_t seg = dh::SegmentId(key_first, key_last, i);
+        return thrust::make_pair(seg, *(val_first + i));
+      });
+  size_t segments_len = key_last - key_first;
+  thrust::fill(thrust::device, key_out, key_out + segments_len, 0);
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/common/device_helpers.cuh]: I think I'm just going to leave it as it's since the bottleneck is in unique by key.

**Comment by trivialfis** [comment]
Added stronger tests on distributed environment.

File: src/common/host_device_vector.cu (hunk 2 above)
[the remark this answers is not in the document]
Comment (RAMitchell) [reply src/common/host_device_vector.cu:208]: Pretty sure it is synchronous unless this document is out of date:
https://developer.nvidia.com/blog/how-overlap-data-transfers-cuda-cc/
"The host memory involved in the data transfer must be pinned memory"

**Pull request merged**